### PR TITLE
Fix the crash issue with objc_sync_enter lock when using Firebase.

### PR DESCRIPTION
**Pull request by pandaleecn**
- Related crashes
https://github.com/firebase/firebase-ios-sdk/issues/5998
https://github.com/firebase/firebase-ios-sdk/issues/4790

- Theoretical reference
https://github.com/apple/darwin-libplatform/blob/main/src/os/lock.c#L494
https://developer.apple.com/documentation/foundation/posixerror/2293610-eownerdead
https://straypixels.net/swift-dictionary-locking/

### CODE REVIEW

File: RxCocoa/Foundation/NSObject+Rx.swift [hunk 1]
```diff
@@ -536,15 +536,31 @@ private extension KeyValueObservingOptions {
...
+	}
+
     func synchronized<T>( _ action: () -> T) -> T {
         objc_sync_enter(self.base)
         let result = action()
         objc_sync_exit(self.base)
         return result
+        lock.lock()
+		defer { lock.unlock() }
+        return action()
```
Comment (freak4pc) [inline RxCocoa/Foundation/NSObject+Rx.swift:563]: Seems you left both the old and new. action() will run twice.
Can you further explain what's the issue ? Add some more context to your fix.

File: RxCocoa/Foundation/NSObject+Rx.swift [hunk 2]
```diff
@@ -536,15 +536,27 @@ private extension KeyValueObservingOptions {
...
 private let deallocSelector = NSSelectorFromString("dealloc")
+private var lockKeyReactive: UInt8 = 0
 
 // MARK: AnyObject + Reactive
 
 extension Reactive where Base: AnyObject {
+
+    private var lock: NSRecursiveLock {
+		get {
+			if let l = objc_getAssociatedObject(self.base, &lockKeyReactive) as? NSRecursiveLock {
```
Comment (freak4pc) [inline RxCocoa/Foundation/NSObject+Rx.swift:547]: I have to say I'm not feeling comfortable creating an associated value like this for each `Reactive` object, especially given the fact we conform many types to it by default. 

Can you share where did you see references that swapping to a lock is the right solution? I saw the issues you linked but not any solutions.

Comment (freak4pc) [reply RxCocoa/Foundation/NSObject+Rx.swift:547]: Should we close this?

**Review by Naveen-C-Ramachandrappa — APPROVED** [review]
Thanks for addressing this crash. Looks good.

**Comment by freak4pc** [comment]
This PR looks quite dangerous IMO and I am not seeing a reasonable explanation it's needed. With no response from the original author, I'm closing this for now.

**Pull request closed without merging**